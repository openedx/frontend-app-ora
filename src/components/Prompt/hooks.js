import { useState } from 'react';

const usePromptHooks = () => {
  const [open, setOpen] = useState(true);

  const toggleOpen = () => setOpen(!open);

  return {
    open,
    toggleOpen,
  };
};

export default usePromptHooks;
