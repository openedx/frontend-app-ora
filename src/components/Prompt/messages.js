import { defineMessages } from '@edx/frontend-platform/i18n';
import { stepNames } from 'constants/index';

const messages = defineMessages({
  [stepNames.submission]: {
    defaultMessage: 'Create a response to the prompt below',
    description: 'Submission step prompt header',
    id: 'frontend-app-ora.Prompt.header.submission',
  },
  [stepNames.studentTraining]: {
    defaultMessage: 'Practice grading a response to the prompt below',
    description: 'Practice step prompt header',
    id: 'frontend-app-ora.Prompt.header.studentTraining',
  },
  [stepNames.self]: {
    defaultMessage: 'Grade your own response to the prompt below',
    description: 'Self step prompt header',
    id: 'frontend-app-ora.Prompt.header.self',
  },
  [stepNames.peer]: {
    defaultMessage: "Grade your peers' responses to the prompt below",
    description: 'Peer step prompt header',
    id: 'frontend-app-ora.Prompt.header.peer',
  },
  [stepNames.done]: {
    defaultMessage: 'Assessment prompt',
    description: 'Done step prompt header',
    id: 'frontend-app-ora.Prompt.header.done',
  },
  promptTitle: {
    defaultMessage: 'Prompt',
    description: 'Prompt title',
    id: 'frontend-app-ora.Prompt.promptTitle',
  },
});

export default messages;
