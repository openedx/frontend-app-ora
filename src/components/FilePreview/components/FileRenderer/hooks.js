import { useState } from 'react';
import { errorStatuses, errorMessages, renderers } from '../constants';
import { getFileType } from '../utils';
import messages from './messages';

/**
 * component hooks
 */
export const useRenderData = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    errorMessages[errorStatus] || errorMessages[errorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = renderers[getFileType(file.fileName)];
  const rendererProps = {
    fileName: file.fileName,
    url: file.fileUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
