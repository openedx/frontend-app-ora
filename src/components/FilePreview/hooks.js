import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';

import {
  PDFRenderer,
  ImageRenderer,
  TXTRenderer,
} from 'components/FilePreview/BaseRenderers';

import messages from './messages';

export const ErrorStatuses = StrictDict({
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
});

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
  txt: 'txt',
  gif: 'gif',
  jfif: 'jfif',
  pjpeg: 'pjpeg',
  pjp: 'pjp',
  svg: 'svg',
});

/**
 * Config data
 */
export const RENDERERS = {
  [FileTypes.pdf]: PDFRenderer,
  [FileTypes.jpg]: ImageRenderer,
  [FileTypes.jpeg]: ImageRenderer,
  [FileTypes.bmp]: ImageRenderer,
  [FileTypes.png]: ImageRenderer,
  [FileTypes.txt]: TXTRenderer,
  [FileTypes.gif]: ImageRenderer,
  [FileTypes.jfif]: ImageRenderer,
  [FileTypes.pjpeg]: ImageRenderer,
  [FileTypes.pjp]: ImageRenderer,
  [FileTypes.svg]: ImageRenderer,
};

export const SUPPORTED_TYPES = Object.keys(RENDERERS);

export const ERROR_STATUSES = {
  [ErrorStatuses.notFound]: messages.fileNotFoundError,
  [ErrorStatuses.serverError]: messages.unknownError,
};

/**
 * Util methods and transforms
 */
export const getFileType = (fileName) => fileName.split('.').pop()?.toLowerCase();
export const isSupported = (file) => SUPPORTED_TYPES.includes(
  getFileType(file.name),
);

export const stateKeys = StrictDict({
  errorStatus: 'errorStatus',
  isLoading: 'isLoading',
});

/**
 * component hooks
 */
export const renderHooks = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useKeyedState(stateKeys.errorStatus, null);
  const [isLoading, setIsLoading] = useKeyedState(stateKeys.isLoading, true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    ERROR_STATUSES[errorStatus] || ERROR_STATUSES[ErrorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = RENDERERS[getFileType(file.name)];
  const rendererProps = {
    fileName: file.name,
    url: file.downloadUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
