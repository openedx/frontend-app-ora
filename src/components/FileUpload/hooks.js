import { useCallback } from 'react';

import { StrictDict, useKeyedState } from '@edx/react-unit-test-utils';
import { useDownloadFiles } from 'hooks/app';

export const stateKeys = StrictDict({
  shouldShowError: 'shouldShowError',
  isModalOpen: 'isModalOpen',
  uploadArgs: 'uploadArgs',
  description: 'description',
});

export const useUploadConfirmModalHooks = ({
  file,
  closeHandler,
  uploadHandler,
}) => {
  const [description, setDescription] = useKeyedState(stateKeys.description, '');
  const [shouldShowError, setShouldShowError] = useKeyedState(
    stateKeys.shouldShowError,
    false,
  );

  const confirmUploadClickHandler = () => {
    // console.log({ confirmUploadClick: { description } });
    if (description !== '') {
      uploadHandler(file, description);
    } else {
      setShouldShowError(true);
    }
  };

  const exitHandler = () => {
    setShouldShowError(false);
    setDescription('');
    closeHandler();
  };

  // Modifying pointer of file object. This is not a good practice.
  // eslint-disable-next-line no-param-reassign, no-return-assign
  const onFileDescriptionChange = (event) => setDescription(event.target.value);

  return {
    shouldShowError,
    confirmUploadClickHandler,
    exitHandler,
    onFileDescriptionChange,
  };
};

export const useFileUploadHooks = ({ onFileUploaded }) => {
  const [uploadArgs, setUploadArgs] = useKeyedState(stateKeys.uploadArgs, {});
  const [isModalOpen, setIsModalOpen] = useKeyedState(
    stateKeys.isModalOpen,
    false,
  );

  const confirmUpload = useCallback(async (file, description) => {
    // console.log({ confirmUpload: { file, description } });
    setIsModalOpen(false);
    if (onFileUploaded) {
      // console.log({ uploadArgs });
      await onFileUploaded({ ...uploadArgs, description });
    }
    setUploadArgs({});
  }, [uploadArgs, onFileUploaded, setIsModalOpen, setUploadArgs]);

  const closeUploadModal = useCallback(() => {
    setIsModalOpen(false);
    setUploadArgs({});
  }, [setIsModalOpen, setUploadArgs]);

  const onProcessUpload = useCallback(
    ({ fileData, handleError, requestConfig }) => {
      setIsModalOpen(true);
      setUploadArgs({ fileData, handleError, requestConfig });
    },
    [setIsModalOpen, setUploadArgs],
  );

  return {
    isModalOpen,
    uploadArgs,
    confirmUpload,
    closeUploadModal,
    onProcessUpload,
  };
};

export const useFileDownloadHooks = ({ files, zipFileName }) => {
  const downloadFileMation = useDownloadFiles();
  const downloadFiles = () => downloadFileMation.mutate({
    files,
    zipFileName,
  });
  return {
    downloadFiles,
    status: downloadFileMation.status,
  };
};

export default {
  useUploadConfirmModalHooks,
  useFileUploadHooks,
  useFileDownloadHooks,
};
