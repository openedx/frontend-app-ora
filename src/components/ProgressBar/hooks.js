import { useParams } from 'react-router-dom';
import { useIsEmbedded, useViewStep } from 'hooks';
import {
  useEffectiveGrade,
  useGlobalState,
} from 'data/services/lms/hooks/selectors';
import {
  stepRoutes,
  stepStates,
} from 'data/services/lms/constants';

export const useProgressStepData = ({ step, canRevisit = false }) => {
  const { xblockId, courseId } = useParams();
  const isEmbedded = useIsEmbedded();
  const viewStep = useViewStep();
  const { stepState } = useGlobalState({ step });

  const href = `/${stepRoutes[step]}${isEmbedded ? '/embedded' : ''}/${courseId}/${xblockId}`;
  const isActive = viewStep === step;
  const isEnabled = (
    isActive
    || (stepState === stepStates.inProgress)
    || (canRevisit && stepState === stepStates.done)
  );
  const myGrade = useEffectiveGrade()?.stepScore;
  return {
    href,
    isEnabled,
    isActive,
    isComplete: stepState === stepStates.done,
    inProgress: stepState === stepStates.inProgress,
    isPastDue: stepState === stepStates.closed,
    myGrade,
    // myGrade: { earned: 8, possible: 10 },
    // isPastDue: step === 'self',
  };
};

export default { useProgressStepData };
