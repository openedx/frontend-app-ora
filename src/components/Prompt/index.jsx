import React from 'react';
import PropTypes from 'prop-types';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Collapsible } from '@edx/paragon';

import { useActiveStepName, useORAConfigData } from 'hooks/app';
import { useViewStep } from 'hooks/routing';

import messages from './messages';

import './index.scss';

const Prompt = ({
  prompt, title, open, onToggle,
}) => {
  const { formatMessage } = useIntl();
  const viewStep = useViewStep();
  const activeStepName = useActiveStepName();
  const message = messages[viewStep] || messages[activeStepName];
  const promptTitle = title || formatMessage(message) || '';
  const imgRegex = /img src="\/asset-v1([^"]*)?"/g;
  const linkRegex = /a href="\/asset-v1([^"]*)?"/g;
  const { baseAssetUrl } = useORAConfigData();
  const promptWithAssets = prompt
    .replaceAll(imgRegex, `img src="${process.env.LMS_BASE_URL}/asset-v1$1"`)
    .replaceAll(linkRegex, `a href="${process.env.LMS_BASE_URL}/asset-v1$1"`);

  const staticRegex = {
    img: /img src="\/static\/([^"]*)?"/g,
    link: /a href="\/static\/([^"]*)?"/g,
  };
  const promptWithStaticAssets = promptWithAssets
    .replaceAll(staticRegex.img, `img src="${process.env.LMS_BASE_URL}/${baseAssetUrl}$1"`)
    .replaceAll(staticRegex.link, `a href="${process.env.LMS_BASE_URL}/${baseAssetUrl}$1"`);

  const collapsibleProps = open !== null && onToggle !== null ? {
    open,
    onToggle,
  } : {
    defaultOpen: true,
  };

  return (
    <Collapsible title={(<h3 className="py-3">{promptTitle}</h3>)} {...collapsibleProps}>
      <div className="prompt" dangerouslySetInnerHTML={{ __html: promptWithStaticAssets }} />
    </Collapsible>
  );
};

Prompt.defaultProps = {
  open: null,
  onToggle: null,
  title: null,
};

Prompt.propTypes = {
  prompt: PropTypes.string.isRequired,
  open: PropTypes.bool,
  onToggle: PropTypes.func,
  title: PropTypes.string,
};

export default Prompt;
