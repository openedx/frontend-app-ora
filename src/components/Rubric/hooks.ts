import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';
import { usePageData, useRubricConfig } from 'data/services/lms/hooks/selectors';
import { submitRubric } from 'data/services/lms/hooks/actions';
import { RubricHookData } from './types';

export const stateKeys = StrictDict({
  rubric: 'rubric',
  overallFeedback: 'overallFeedback',
});

export const useRubricData = ({ isGrading }): RubricHookData => {
  const data = usePageData();
  const { criteria, feedbackConfig } = useRubricConfig();

  const [rubricData, setRubricData] = useKeyedState(stateKeys.rubric, data.rubric);
  const [overallFeedback, setOverallFeedback] = useKeyedState(
    stateKeys.overallFeedback, data.rubric.overallFeedback
  );
  const submitRubricMutation = submitRubric();

  const onOverallFeedbackChange = (e: React.ChangeEvent<HTMLInputElement>) =>
    setOverallFeedback(e.target.value);

  const onSubmit = () => {
    submitRubricMutation.mutate({
      overallFeedback,
      optionsSelected: rubricData.optionsSelected,
      criterionFeedback: rubricData.criterionFeedback,
    } as any);
  };

  return {
    rubricData,
    setRubricData,
    criteria: criteria.map((criterion) => ({
      ...criterion,
      optionsValue: rubricData.optionsSelected[criterion.name],
      optionsIsInvalid: false,
      optionsOnChange: (e: React.ChangeEvent<HTMLInputElement>) => {
        setRubricData({
          ...rubricData,
          optionsSelected: {
            ...rubricData.optionsSelected,
            [criterion.name]: e.target.value,
          },
        });
      },

      feedbackValue: rubricData.criterionFeedback[criterion.name],
      feedbackIsInvalid: false,
      feedbackOnChange: (e: React.ChangeEvent<HTMLInputElement>) => {
        setRubricData({
          ...rubricData,
          criterionFeedback: {
            ...rubricData.criterionFeedback,
            [criterion.name]: e.target.value,
          },
        });
      },
    })),
    onSubmit,
    submitStatus: submitRubricMutation.status,

    // overall feedback
    overallFeedback,
    onOverallFeedbackChange,
    overallFeedbackDisabled: !isGrading,
    overallFeedbackIsInvalid: false,
    overallFeedbackPrompt: feedbackConfig.defaultText,
  };
};

export default useRubricData;
