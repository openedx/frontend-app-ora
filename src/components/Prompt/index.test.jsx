import { shallow } from '@edx/react-unit-test-utils';

import { useActiveStepName, useORAConfigData } from 'hooks/app';
import { useViewStep } from 'hooks/routing';

import Prompt from './index';

jest.mock('hooks/app', () => ({
  useActiveStepName: jest.fn(),
  useORAConfigData: jest.fn(),
}));
jest.mock('hooks/routing', () => ({
  useViewStep: jest.fn(),
}));

describe('<Prompt />', () => {
  const props = {
    prompt: 'prompt',
    title: 'title',
  };

  useActiveStepName.mockReturnValue('activeStepName');
  useORAConfigData.mockReturnValue({ baseAssetUrl: 'baseAssetUrl/' });
  useViewStep.mockReturnValue('viewStep');

  it('render default', () => {
    const wrapper = shallow(<Prompt {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    // if open and onToggle are not passed, defaultOpen should be true
    expect(wrapper.instance.findByType('Collapsible')[0].props.defaultOpen).toBe(true);
  });

  it('render with open and onToggle', () => {
    const wrapper = shallow(<Prompt {...props} open onToggle={() => {}} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    // if open and onToggle are passed, open and onToggle should be passed to Collapsible
    expect(wrapper.instance.findByType('Collapsible')[0].props.open).toBe(true);
    expect(wrapper.instance.findByType('Collapsible')[0].props.onToggle).toBeInstanceOf(Function);
    expect(wrapper.instance.findByType('Collapsible')[0].props.defaultOpen).toBe(undefined);
  });

  it('render without title', () => {
    const wrapper = shallow(<Prompt {...props} title={null} />);
    expect(wrapper.snapshot).toMatchSnapshot();
  });

  describe('test prompt', () => {
    const getPromptHtml = (prompt) => {
      const wrapper = shallow(<Prompt {...props} prompt={prompt} />);
      // eslint-disable-next-line no-underscore-dangle
      return wrapper.instance.findByType('div')[0].props.dangerouslySetInnerHTML.__html;
    };
    const abitraryEndpoint = '/abc/def/ghi';
    const fullAssetUrl = `http://localhost:18000/asset-v1${abitraryEndpoint}`;
    const fullStaticAssetUrl = `http://localhost:18000/baseAssetUrl${abitraryEndpoint}`;
    const relativeAssetUrl = `/asset-v1${abitraryEndpoint}`;
    const relativeStaticAssetUrl = `/static${abitraryEndpoint}`;

    it('does not update url for anchor and image that is not using relative url', () => {
      expect(getPromptHtml(`<img src="${fullAssetUrl}" />`)).toBe(`<img src="${fullAssetUrl}" />`);
      expect(getPromptHtml(`<a href="${fullAssetUrl}" />`)).toBe(`<a href="${fullAssetUrl}" />`);
      // ignore non image and anchor
      expect(getPromptHtml(`<object data="${relativeAssetUrl}" />`)).toBe(`<object data="${relativeAssetUrl}" />`);
    });

    it('update assets url for anchor and image', () => {
      expect(getPromptHtml(`<img src="${relativeAssetUrl}" />`)).toBe(`<img src="${fullAssetUrl}" />`);
      expect(getPromptHtml(`<a href="${relativeAssetUrl}" />`)).toBe(`<a href="${fullAssetUrl}" />`);
    });

    it('update static assets url for anchor and image', () => {
      expect(getPromptHtml(`<img src="${relativeStaticAssetUrl}" />`)).toBe(`<img src="${fullStaticAssetUrl}" />`);
      expect(getPromptHtml(`<a href="${relativeStaticAssetUrl}" />`)).toBe(`<a href="${fullStaticAssetUrl}" />`);
    });
  });
});
