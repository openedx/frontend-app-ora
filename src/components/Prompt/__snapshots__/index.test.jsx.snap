// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Prompt /> render default 1`] = `
<Collapsible
  defaultOpen={true}
  title={
    <h3
      className="py-3"
    >
      title
    </h3>
  }
>
  <div
    className="prompt"
    dangerouslySetInnerHTML={
      Object {
        "__html": "prompt",
      }
    }
  />
</Collapsible>
`;

exports[`<Prompt /> render with open and onToggle 1`] = `
<Collapsible
  onToggle={[Function]}
  open={true}
  title={
    <h3
      className="py-3"
    >
      title
    </h3>
  }
>
  <div
    className="prompt"
    dangerouslySetInnerHTML={
      Object {
        "__html": "prompt",
      }
    }
  />
</Collapsible>
`;

exports[`<Prompt /> render without title 1`] = `
<Collapsible
  defaultOpen={true}
  title={
    <h3
      className="py-3"
    >
      Prompt
    </h3>
  }
>
  <div
    className="prompt"
    dangerouslySetInnerHTML={
      Object {
        "__html": "prompt",
      }
    }
  />
</Collapsible>
`;
